include/revng/ABI/CallingConventionTrait.h

```diff
@@ -0,0 +1,1083 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "revng/Model/Binary.h"
+#include "revng/Model/Register.h"
+#include "revng/Model/Type.h"
+
+namespace abi {
+
+/// A trait struct used for describing the specifics of the calling convention
+/// for an ABI relevant for revng model.
+///
+/// To register a new ABI for use with the `abi::ABI` template, add
+/// a specialization of this trait.
+template<model::abi::Values V>
+struct CallingConventionTrait {
+
+  /// States whether the ABI is focused on preserving the argument order.
+  ///
+  /// Here's an example function:
+  /// ```
+  /// struct Big; // Is really big, so it can only be passed in memory.
+  /// void function(Big, signed, float, unsigned);
+  /// ```
+  ///
+  /// If the ABI is position based, the arguments would be passed in
+  /// - memory for the `Big` argument (and a pointer to it in the first GRP).
+  /// - the second GPR register for the `signed` argument.
+  /// - the third vector register for the `float` argument.
+  /// - the forth GPR register for the `unsigned` argument.
+  ///
+  /// If the ABI is NOT position based, the arguments would be passed in
+  /// - memory for the `Big` argument.
+  /// - the first GPR register for the `signed` argument.
+  /// - the first vector register for the `float` argument.
+  /// - the second GPR register for the `unsigned` argument.
+  ///
+  /// A typical example of a position-based ABI is `Microsoft_x86_64`,
+  /// a non-position-based one - `SystemV_x86_64`.
+  static constexpr bool ArgumentsArePositionBased = false;
+
+  /// States whether an object that needs two GPRs to fit (5-8 bytes on 32-bit
+  /// architectures and 9-16 bytes on 64-bit systems) are only allowed to start
+  /// from a register with an even index.
+  ///
+  /// Here's an example function:
+  /// ```
+  /// void function(uint32_t, uint64_t);
+  /// ```
+  /// On a system with 32-bit wide GPRs, the first argument (`uint32_t`) is
+  /// passed using the first allowed GPR (say `r0`).
+  ///
+  /// The second argument (`uint64_t`) requires two register to fit, so it's
+  /// passed using second and third registers if the ABI allows starting double
+  /// arguments from any register (`r1` and `r2` in this example), or third and
+  /// forth if it only allows starting them from even registers (`r2` and `r3`
+  /// in this example, since `r1` is considered an odd register (the count
+  /// starts from 0, much like C array indexing).
+  ///
+  /// \note: this option is only applicable for non-position based ABIs
+  /// (if `ArgumentsArePositionBased` is `false`).
```

Comment:
I want some `static_assert` for this kind of stuff somewhere.

Reply:
What do you mean? If `ArgumentsArePositionBased == true`, whatever the value of this parameter is - it doesn't matter. And if `ArgumentsArePositionBased == false` - the value changes the behaviour. How do you `assert` the relation like that?

Reply:
Mmmh, right. We could assert that the value has not been "customized", but yeah, not very helpful.